### DISCUSSION TITLE
[ Vue ] TypeScript error on `props` when passed to `mount`

### DISCUSSION THREAD

**Discussion by jon-karlsen** [cypress-io/cypress#25563] [Questions and Help]
Setup based on [this example](https://docs.cypress.io/guides/component-testing/vue/examples#Vue-Router) produces the following error for each occurrence `props` being passed to `mount`:

```ts
Type '{ batchOpsConfig: __Config; }' is not assignable to type 'VNodeProps & { __v_isVNode?: undefined; [Symbol.iterator]?: undefined; } & Record<string, any> & { [x: number]: unknown; } & { readonly length?: number | Prop<unknown, unknown> | null | undefined; ... 22 more ...; toLocaleString?: string | undefined; } & AllowedComponentProps & ComponentCustomProps'.
  Type '{ batchOpsConfig: __Config; }' is not assignable to type '{ readonly length?: number | Prop<unknown, unknown> | null | undefined; readonly concat?: Prop<unknown, unknown> | { (...items: ConcatArray<string>[]): string[]; (...items: (string | ConcatArray<string>)[]): string[]; } | null | undefined; ... 21 more ...; toLocaleString?: string | undefined; }'.
    Types of property 'toString' are incompatible.
      Type '() => string' is not assignable to type 'string'.
```

```ts
// example.cy.ts
describe( '<HeaderLinkComponent />' , () => {

    it( 'renders' , () => {
        const displayText = 'TestLink'
            , linkPath    = '/test/path'

        cy.mount( HeaderLinkComponent 
                , { 
                    props : {
                        displayText 
                      , linkPath 
                    } 
                } 
            )

        cy.contains( displayText )

        cy.get( 'a' )
          .invoke( 'attr' , 'href' )
          .should( 'eq' , linkPath )
    } )

} )
```

While all relevant tests do work, and I can suppress the error using `//@ts-ignore`, I'd obviously prefer to type `mount` properly. I can't find any information on how to do so in the docs or any relevant issues to work off of. I've messed around a little with various types and combinations thereof, but it's not making me much the wiser--any ideas on how to go about fixing this?

```ts
// component.ts

type MountParams  = Parameters<typeof mount>
type OptionsParam = MountParams[ 1 ] 
                    & { router? : Router }


declare global {
   // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
           /**
            *      Helper mount function for Vue Components
            * 
            *      @param component    Vue Component or JSX Element to mount
            *      @param options      Options passed to Vue Test Utils
            * 
            */
            mount( component : any , options? : OptionsParam ) : Chainable<any>
        }
    }
}


Cypress.Commands.add( 'mount' , ( component : any , options : OptionsParam = {} ) => {
    options.global         = options.global         || {}
    options.global.plugins = options.global.plugins || []
  
    if ( !options.router ) 
        options.router = createRouter( {
            routes  : routes
          , history : createMemoryHistory()
            } )

    options.global.plugins.push( {
        install( app ) {
            app.use( options.router as never )
            app.use( i18n )
            app.use( FloatingVue )
            app.use( Toast, toastOpt )
            }
        } )
  
    return mount( component , options )

    } )
```

Using:
- `vue`: 3.2.45
- `cypress` : 12.4.0

**Comment by portyu9** [comment]
Hi @jon-karlsen 

The strange `length`, `concat`, and `toString` properties are coming from the **typing of your custom `mount` command**, not from the props you are passing.

`mount()` has a number of generic/overloaded signatures for the different forms a Vue component can take. This:

```ts
type MountParams = Parameters<typeof mount>
type OptionsParam = MountParams[1]
```

extracts one parameter representation from that overloaded function, and then this declaration:

```ts
mount(component: any, options?: OptionsParam): Chainable<any>
```

loses the relationship between the actual component and its props.

That is why TypeScript ends up checking your props against one of the wrong Vue component shapes and you get array-like properties such as `length`, `concat`, etc.

If possible, keep the public `cy.mount()` type identical to Cypress's real `mount()` function:

```ts
import { mount } from 'cypress/vue'

declare global {
  namespace Cypress {
    interface Chainable {
      mount: typeof mount
    }
  }
}
```

Then keep the router setup as implementation detail of your custom command:

```ts
Cypress.Commands.add('mount', (component, options = {}) => {
  options.global = options.global || {}
  options.global.plugins = options.global.plugins || []

  const router = createRouter({
    routes,
    history: createMemoryHistory(),
  })

  options.global.plugins.push({
    install(app) {
      app.use(router)
      app.use(i18n)
      app.use(FloatingVue)
      app.use(Toast, toastOpt)
    },
  })

  return mount(component, options)
})
```

This preserves Cypress's own overloaded `mount` signatures, including the inference from the `.vue` component to the allowed `props`.

If you genuinely need callers to supply a custom router as an extra option, that becomes harder to type without weakening the component-specific prop inference. In that case I would either:

- expose the router through a separate custom command/argument, or
- type the extended options with `CyMountOptions<any> & { router?: Router }`, accepting that the custom extension gives up some of the precise per-component prop inference.

For the example shown in the question, keeping `mount: typeof mount` and creating the router inside the command is the cleanest fix.

Please mark this answer as accepted if it helped, thank you.
